### PR TITLE
feat: added tiles api for tiles count spec

### PR DESCRIPTION
**Pull request by CL-SHLOMIKONCHA**
| Question                | Answer                                                                          |
| ---------------- | -------------------------------------------------------------------------- |
| Bug fix         | ✖                                                                        |
| New feature     | ✔                                                                        |
| Breaking change | ✖                                                                        |
| Deprecations    | ✖                                                                        |
| Documentation   | ✖                                                                        |
| Tests added     |✖                                                                        |
| Chore            | ✔                                                                       |

Related issues: #XXX , #XXX ...
Closes #XXX ...

Further  information:

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,5 +1,4 @@
-# Map Colonies typescript service template
-
+# layer-spec
```
Comment (asafMasa) [inline README.md:1]: add detailed explanation about the service usage and how to configure it

File: config/custom-environment-variables.json [hunk 2]
```diff
@@ -38,5 +38,22 @@
         }
       }
     }
+  },
+  "db": {
+    "host": "DB_HOST",
+    "port": "DB_PORT",
```
Comment (asafMasa) [inline config/custom-environment-variables.json]: add "__format": "XXX"

File: config/custom-environment-variables.json [hunk 3]
```diff
@@ -38,5 +38,22 @@
...
+    "user": "DB_USER",
+    "password": "DB_PASSWORD",
+    "database": "DB_NAME",
+    "enableSslAuth": {
+      "__name": "DB_SSL_ENABLE",
+      "__format": "boolean"
+    },
+    "rejectUnauthorized": "DB_SSL_REJECT_UNAUTHORIZED",
```
Comment (asafMasa) [inline config/custom-environment-variables.json]: add "__format": "XXX"

File: openapi3.yaml [hunk 4]
```diff
@@ -1,51 +1,63 @@
 openapi: 3.0.1
 info:
-  title: service-name
-  description: basic template for map colonies service
+  title: Sync Worker
```
Comment (asafMasa) [inline openapi3.yaml]: layer-spec

File: openapi3.yaml [hunk 5]
```diff
@@ -1,51 +1,63 @@
...
+  description: queryable database service for layer specification
```
Comment (asafMasa) [inline openapi3.yaml]: fix description

File: package.json [hunk 6]
```diff
@@ -1,11 +1,11 @@
...
+  "name": "layer-spec",
   "version": "1.0.0",
-  "description": "This is template for map colonies typescript service",
+  "description": "queryable database service for layer specification ",
   "main": "./src/index.ts",
   "scripts": {
-    "test:unit": "jest --config=./tests/configurations/unit/jest.config.js",
-    "test:integration": "jest --config=./tests/configurations/integration/jest.config.js",
+    "test:unit": "jest --config=./tests/configurations/unit/jest.config.js --passWithNoTests",
+    "test:integration": "jest --config=./tests/configurations/integration/jest.config.js --passWithNoTests",
```
Comment (asafMasa) [inline package.json]: passWithNoTests - remove and add tests

File: src/tiles/controllers/tilesController.ts [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+@injectable()
+export class TilesController {
+  private readonly createdResourceCounter: BoundCounter;
+
+  public constructor(
+    @inject(Services.LOGGER) private readonly logger: Logger,
+    @inject(TilesManager) private readonly manager: TilesManager,
+    @inject(Services.METER) private readonly meter: Meter
+  ) {
+    this.createdResourceCounter = meter.createCounter('created_resource');
```
Comment (asafMasa) [inline src/tiles/controllers/tilesController.ts]: created_resource - change name

File: src/tiles/models/tilesManager.ts [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+    @inject(Services.DB_CONFIG) private readonly dbConfig: IDBConfig,
+    private readonly pgClient: PgClient
+  ) {
+    this.logger = logger;
+  }
+  public async getTilesCount(layerId: string): Promise<ITilesCountResponse> {
+    try {
+      const query = `SELECT "tilesCount" FROM "TilesCounter" WHERE "layerId"='${layerId}'`;
+      const result = await this.pgClient.execute<ITilesCountResponse>(query);
+      const data = result[0];
```
Comment (asafMasa) [inline src/tiles/models/tilesManager.ts]: if the layer doesn't exist, what is the returned value from the pgClient? is it 0, undefined or throws an error?
In the end you should throw a NotFound error

File: src/tiles/models/tilesManager.ts [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+    }
+  }
+
+  public async upsertTilesCount(layerId: string, tilesBatchCount: number): Promise<void> {
+    try {
+      const query = `INSERT INTO "TilesCounter" ("tilesCount", "layerId") 
+      VALUES (${tilesBatchCount}, '${layerId}') 
+      ON CONFLICT ("layerId") DO UPDATE SET "tilesCount" = "TilesCounter"."tilesCount" + ${tilesBatchCount}`;
+      await this.pgClient.execute(query);
+      this.logger.info(`tiles batch count for layerId: '${layerId}' updated successfully`);
```
Comment (asafMasa) [inline src/tiles/models/tilesManager.ts]: move to before the pg execution

**Review by asafMasa — CHANGES_REQUESTED** [review]
🥇 Good job

File: src/tiles/models/tilesManager.ts (hunk 8 above)
Comment (CL-SHLOMIKONCHA) [reply src/tiles/models/tilesManager.ts]: fixed in e438346

File: package.json (hunk 6 above)
Comment (CL-SHLOMIKONCHA) [reply package.json]: fixed in e438346

File: openapi3.yaml (hunk 5 above)
Comment (CL-SHLOMIKONCHA) [reply openapi3.yaml]: fixed in e438346

File: openapi3.yaml (hunk 4 above)
Comment (CL-SHLOMIKONCHA) [reply openapi3.yaml]: fixed in e438346

File: config/custom-environment-variables.json (hunk 3 above)
Comment (CL-SHLOMIKONCHA) [reply config/custom-environment-variables.json]: fixed in e438346

File: config/custom-environment-variables.json (hunk 2 above)
Comment (CL-SHLOMIKONCHA) [reply config/custom-environment-variables.json]: fixed in e438346

File: README.md (hunk 1 above)
Comment (CL-SHLOMIKONCHA) [reply README.md:1]: fixed in 487f1ba

File: src/tiles/controllers/tilesController.ts (hunk 7 above)
Comment (CL-SHLOMIKONCHA) [reply src/tiles/controllers/tilesController.ts]: fixed in 76fb305

File: src/tiles/models/tilesManager.ts (hunk 9 above)
Comment (CL-SHLOMIKONCHA) [reply src/tiles/models/tilesManager.ts]: fixed in  ef63b3c

File: README.md [hunk 10]
```diff
@@ -119,3 +112,34 @@ To only run integration tests:
...
 npm run test:integration
 ```
+
+## Configurations
+
+ENVS:
+
+SERVER_PORT set the server port number - deafult to 8080
+
+LOG_LEVEL set the log level *based on 'winston' logger, available values as declared in winston logger docs, default to 'info'
```
Comment (asafMasa) [inline README.md]: @map-colonies/js-logger is not 'winston'

File: README.md [hunk 11]
```diff
@@ -1,15 +1,8 @@
...
-![badge-alerts-lgtm](https://img.shields.io/lgtm/alerts/github/MapColonies/ts-server-boilerplate?style=for-the-badge)
 
-![grade-badge-lgtm](https://img.shields.io/lgtm/grade/javascript/github/MapColonies/ts-server-boilerplate?style=for-the-badge)
-
-![snyk](https://img.shields.io/snyk/vulnerabilities/github/MapColonies/ts-server-boilerplate?style=for-the-badge)
-
-----------------------------------
-
-This is a basic repo template for building new MapColonies web services in Typescript.
+This service exposes CRUD (Create, Read, Update and Delete) operations into/from layer-spec database.
```
Comment (asafMasa) [inline README.md:5]: add a line on the tiles count

File: src/tiles/models/tilesManager.ts [hunk 12]
```diff
@@ -15,33 +15,27 @@ export interface ITilesCountRequest {
...
-      await this.pgClient.execute(query);
-      this.logger.info(`tiles batch count for layerId: '${layerId}' updated successfully`);
-    } catch (error) {
-      throw new Error(error);
-    }
+    const query = `INSERT INTO "TilesCounter" ("tilesCount", "layerId") 
+      VALUES ($1, $2) 
+      ON CONFLICT ("layerId") DO UPDATE SET "tilesCount" = "TilesCounter"."tilesCount" + $1`;
+    const values = [tilesBatchCount, layerId];
+    this.logger.info(`updating tiles batch count for layerId: '${layerId}' in database`);
```
Comment (asafMasa) [inline src/tiles/models/tilesManager.ts]: log at the start of the method

**Review by asafMasa — CHANGES_REQUESTED** [review]

File: src/tiles/models/tilesManager.ts (hunk 12 above)
Comment (CL-SHLOMIKONCHA) [reply src/tiles/models/tilesManager.ts]: fixed it aa489ae

File: README.md (hunk 11 above)
Comment (CL-SHLOMIKONCHA) [reply README.md:5]: fixed it aa489ae

File: README.md (hunk 10 above)
Comment (CL-SHLOMIKONCHA) [reply README.md]: fixed it aa489ae

**Review by asafMasa — APPROVED** [review]

**Pull request merged**